core/src/main/scala/epollcat/internal/ch/SocketHelpers.scala

```diff
@@ -120,6 +136,56 @@ private[ch] object SocketHelpers {
     new InetSocketAddress(inetAddr, port)
   }
 
+  def toInet6SocketAddress(
+      addr: Ptr[posix.netinet.in.sockaddr_in6]
+  ): InetSocketAddress = {
+    val port = posix.arpa.inet.htons(addr.sin6_port).toInt
+    val addrBytes = addr.sin6_addr.at1.asInstanceOf[Ptr[Byte]]
+    val inetAddr = InetAddress.getByAddress {
+      val addr = new Array[Byte](16)
+      var i = 0
+      while (i < addr.length) {
+        addr(i) = addrBytes(i.toLong)
+        i += 1
+      }
+      addr
+    }
+    new InetSocketAddress(inetAddr, port)
+  }
+
+  def toAddrinfo(addr: InetSocketAddress): Either[Throwable, Ptr[posix.netdb.addrinfo]] = Zone {
+    implicit z =>
+      val addrinfo = stackalloc[Ptr[posix.netdb.addrinfo]]()
+      val hints = stackalloc[posix.netdb.addrinfo]()
+      hints.ai_family =
+        if (preferIPv4Stack)
+          posix.sys.socket.AF_INET
+        else
+          posix.sys.socket.AF_INET6
+      hints.ai_flags = posix.netdb.AI_NUMERICHOST | posix.netdb.AI_NUMERICSERV
+      if (!preferIPv4Stack) hints.ai_flags |= posix.netdb.AI_V4MAPPED
```

Comment:
I am pretty sure this wants to be 
`(posix.netdb.AI_V4MAPPED | posix.netdb.AI_ADDRCONFIG)`

That is, if you naturally have no IPv6 on the system, do not
take a prefectly good IPv4 address and map it to a useless
IPv4mappedIPv6 address :smile:

Gotta love the wide variety of cases we are trying to cover.

Reply:
This is another one that I'm confused about. The problem is, we've already created the socket according to the value of `preferIPv4Stack`. So it seems to me that using `AI_ADDRCONFIG` could get us into trouble?

In the specific case you mention, if `preferIPv4Stack = false` and the system has no IPv6, then we are already hosed regardless of our choice here 😅

Reply:
I think the difference is between `java.net.preferIPv4Stack` which can and usually is false,
and the variable preferIPv4Stack.  The latter is the reason for the startup runtime test
to see if IPv6 is available.  If there is no IPv6, I believe that would set the variable 
to true, even if the property is false. The property is "prefer", like a hint.

This is reasoning from the fact that the system property defaults to false,
yet Java appears to run just fine on IPv4 only systems (if you can find one).

My head is spinning.

Reply:
> If there is no IPv6, I believe that would set the variable
to true, even if the property is false.

Yes, I agree. In fact, when we implement this, we should really introduce a new variable `useIPv4Stack` that is derived from the combination of `preferIPv4Stack` with the startup runtime test.

Reply:
Agreed. I thought of that for SN 0.5.0 and do not remember what I did. If I left it as `preferIPv4` it is
becoming evident that was a mistake.

Reply:
For historical reference, as we move towards closure:

In tracing this yet again, the line
`if (!preferIPv4Stack) hints.ai_flags |= posix.netdb.AI_V4MAPPED`
is directly equivalent to OR'in in AI_ADDRCONFIG.

My eye is just trained to expect (AI_V4MAPPED | AI_ADDRCONFIG)
and alerted.

No need for the OR in this case.